apps/hash-ai-worker-ts/src/activities/flow-activities/shared/get-file-entity-from-url.ts

```diff
@@ -0,0 +1,273 @@
+import { generateUuid } from "@local/hash-isomorphic-utils/generate-uuid";
+import { createDefaultAuthorizationRelationships } from "@local/hash-isomorphic-utils/graph-queries";
+import { systemEntityTypes } from "@local/hash-isomorphic-utils/ontology-type-ids";
+import { mapGraphApiEntityMetadataToMetadata } from "@local/hash-isomorphic-utils/subgraph-mapping";
+import type { FileProperties } from "@local/hash-isomorphic-utils/system-types/shared";
+import type { EntityMetadata } from "@local/hash-subgraph";
+import mime from "mime-types";
+
+import { getFlowContext } from "../../shared/get-flow-context";
+import { graphApiClient } from "../../shared/graph-api-client";
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = dirname(__filename);
+
+const baseFilePath = join(__dirname, "/var/tmp_files");
+
+const downloadFileToFileSystem = async (fileUrl: string) => {
+  mkdirSync(baseFilePath, { recursive: true });
+
+  const tempFileName = generateUuid();
+  const filePath = join(baseFilePath, tempFileName);
+
+  const response = await fetch(fileUrl);
+  if (!response.ok || !response.body) {
+    throw new Error(`Failed to fetch ${fileUrl}: ${response.statusText}`);
+  }
+
+  try {
+    const fileStream = createWriteStream(filePath);
+    await finished(
+      Readable.fromWeb(response.body as ReadableStream<Uint8Array>).pipe(
+        fileStream,
+      ),
+    );
+    return filePath;
+  } catch (error) {
+    await unlink(filePath);
+    throw new Error(
+      `Failed to write file to file system: ${(error as Error).message}`,
+    );
+  }
+};
+
+const writeFileToS3URL = async ({
+  filePath,
+  mimeType,
+  presignedPutUrl,
+  sizeInBytes,
+}: {
+  filePath: string;
+  mimeType: string;
+  presignedPutUrl: string;
+  sizeInBytes: number;
+}) => {
+  return new Promise((resolve, reject) => {
+    const fileStream = createReadStream(filePath);
+    const req = (presignedPutUrl.startsWith("https://") ? https : http).request(
+      presignedPutUrl,
+      {
+        headers: { "Content-Length": sizeInBytes, "Content-Type": mimeType },
+        method: "PUT",
+      },
+      (res) => {
+        res.on("data", () => {});
+        res.on("end", () => {
+          if (res.statusCode && res.statusCode >= 200 && res.statusCode < 300) {
+            resolve("Ok");
+          } else {
+            reject(
+              new Error(
+                `${res.statusCode} Error uploading to S3: ${res.statusMessage}`,
+              ),
+            );
+          }
+        });
+      },
+    );
+
+    req.on("error", (error) => {
+      reject(new Error(`Request error: ${error.message}`));
+    });
+
+    fileStream.on("error", (error) =>
+      reject(new Error(`Error reading file: ${error.message}`)),
+    );
+    fileStream.pipe(req);
+    fileStream.on("end", () => req.end());
+  }).finally(() => {
+    void unlink(filePath);
+  });
+};
+
+export const getFileEntityFromUrl = async (params: {
```

Comment:
Can we require provenance here now (which may be empty) and pass it where we call it?

Reply:
What kind of provenance information are you referring to?

Reply:
Whatever we can populate – the edition-level metadata at least (e.g. the fact that it's a flow, and the flow id)

When we have a description / title inferred from the AI we can also add that. Maybe the webpage where it was found, for a source on the file URL or something?

Reply:
I've added the origin to the file entity in https://github.com/hashintel/hash/pull/4399/commits/6a0e5ffd6e5c0e9b020179ab32acc59e3e1352f1, but haven't specified the `actorType` because getting a file from a URL isn't really an "AI" - perhaps we need an `"Machine"` value for this field or something like that?

> When we have a description / title inferred from the AI we can also add that. Maybe the webpage where it was found, for a source on the file URL or something?

I suppose we could do this for the edition, since we have a `sources` array there, but wouldn't these fields make more sense as properties on the file entity, rather than metadata in the provenance?

Reply:
Yes, a `@todo` would be fine. Presumably we have it where we propose the file entity though, can we add it there as a source the same way we do for other proposed entiites?

Reply:
I've added a `@todo`, and accounted for metadata set on the proposed file entity in https://github.com/hashintel/hash/pull/4399/commits/bc016a28eb42b53c282ba5f7c1b0e71c3557bebf